pyschema_extensions/avro_schema_parser.py

```diff
@@ -43,118 +43,121 @@ def parse_schema_string(schema_string):
+            field_name = field_def["name"]
+            field_builder = self._get_field_builder(field_def["type"])
+
+            if "default" in field_def:
+                if field_def["default"] is not None:
+                    default_parser = field_builder(nullable=False).avro_load
+                    default_value = default_parser(field_def["default"])
+                else:
+                    default_value = None
             else:
-                default_value = None
+                default_value = pyschema.core.NO_DEFAULT
+
+            field = field_builder(
+                description=field_def.get("doc"),
+                default=default_value
+            )
+            field_dct[field_name] = field
+            field_dct["__module__"] = "__avro_parser_runtime__"  # not great, but better than "abc"
+            if "namespace" in schema_struct:
+                field_dct["_namespace"] = schema_struct["namespace"]
+
+        if "doc" in schema_struct:
+            field_dct["__doc__"] = schema_struct["doc"]
+        schema = pyschema.core.PySchema(record_name.encode("ascii"), (pyschema.core.Record,), field_dct)
+        return schema
+
+    def _get_field_builder(self, type_def_struct):
+        if isinstance(type_def_struct, list):
+            return self._parse_union(type_def_struct)
+        elif isinstance(type_def_struct, dict):
+            return self._parse_complex(type_def_struct)
+        elif type_def_struct in SIMPLE_FIELD_MAP:
+            field_builder = SIMPLE_FIELD_MAP.get(type_def_struct)
+            return partial(field_builder, nullable=False)
         else:
-            default_value = pyschema.core.NO_DEFAULT
+            # Default case is that we reference an already defined sub record
+            return self._parse_subrecord(type_def_struct)
+
+    def _parse_union(self, union_struct):
+        filtered = [subtype for subtype in union_struct if subtype != "null"]
+        if len(filtered) > 1:
+            raise AVSCParseException(
+                "Unions with multiple non-null as type are unsupported: {0}".format(union_struct)
+            )
+        nullable = "null" in union_struct
+        actual_type = filtered[0]
+        field_builder = self._get_field_builder(actual_type)
+
+        return partial(field_builder, nullable=nullable)
+
+    def _parse_map(self, type_def_struct):
+        values_field_builder = self._get_field_builder(type_def_struct["values"])
+        return partial(
+            pyschema.Map,
+            # NOTE: values field builder is executed here, so the field index
+            # will be created at this point, but that shouldn't matter for the map type
+            value_type=values_field_builder(),
+            nullable=False
+        )
 
-        field = field_builder(
-            description=field_def.get("doc"),
-            default=default_value
+    def _parse_subrecord(self, type_def_struct):
+        if "fields" in type_def_struct:
```

Comment:
There should be a check here to see if type_def_struct is a dictionary as well. Otherwise, there will be a very funky bug if anyone declares a record with a name including the substring "fields" :)